src/map/entities/mobentity.cpp

```diff
@@ -803,150 +803,160 @@ void CMobEntity::DropItems()
 
-            DropList_t* DropList = itemutils::GetDropList(m_DropID);
-            //ShowDebug(CL_CYAN"DropID: %u dropping with TH Level: %u\n" CL_RESET, PMob->m_DropID, PMob->m_THLvl);
-
-            if (DropList != nullptr && !getMobMod(MOBMOD_NO_DROPS) && DropList->Items.size() || DropList->Groups.size())
+            // check for gil (beastmen drop gil, some NMs drop gil)
+            if (CanDropGil() || (map_config.all_mobs_gil_bonus > 0 && getMobMod(MOBMOD_GIL_MAX) >= 0)) // Negative value of MOBMOD_GIL_MAX is used to prevent gil drops in Dynamis/Limbus.
             {
-                //THLvl is the number of 'extra chances' at an item. If the item is obtained, then break out.
-                uint8 maxRolls = 1 + (m_THLvl > 2 ? 2 : m_THLvl);
-                uint8 bonus = (m_THLvl > 2 ? (m_THLvl - 2) * 10 : 0);
+                charutils::DistributeGil(PChar, this); // TODO: REALISATION MUST BE IN TREASUREPOOL
+            }
 
-                for (const DropGroup_t&  group : DropList->Groups)
-                {
-                    for (uint8 roll = 0; roll < maxRolls; ++roll)
-                    {
-                        //Determine if this group should drop an item
-                        if (group.GroupRate > 0 && dsprand::GetRandomNumber(1000) < group.GroupRate * map_config.drop_rate_multiplier + bonus)
-                        {
-                            //Each item in the group is given its own weight range which is the previous value to the previous value + item.DropRate
-                            //Such as 2 items with drop rates of 200 and 800 would be 0-199 and 200-999 respectively
-                            uint16 previousRateValue = 0;
-                            uint16 itemRoll = dsprand::GetRandomNumber(1000);
-                            for (const DropItem_t& item : group.Items)
-                            {
-                                if (previousRateValue + item.DropRate > itemRoll)
-                                {
-                                    PChar->PTreasurePool->AddItem(item.ItemID, this);
-                                    break;
-                                }
-                                previousRateValue += item.DropRate;
-                            }
-                            break;
-                        }
-                    }
-                }
+            DropItems(PChar);
+        }
+
+        PChar->setWeaponSkillKill(false);
+        StatusEffectContainer->KillAllStatusEffect();
+
+        // NOTE: this is called for all alliance / party members!
+        luautils::OnMobDeath(this, PChar);
+
+    }
+    else
+    {
+        luautils::OnMobDeath(this, nullptr);
+    }
+}
+
+void CMobEntity::DropItems(CCharEntity* PChar) {
+    //Limit number of items that can drop to the treasure pool size
+    uint8 dropCount = 0;
+    #define ADD_ITEM_TO_POOL(ItemID) { PChar->PTreasurePool->AddItem(ItemID, this); if( ++dropCount == TREASUREPOOL_SIZE ) { return; } }
```

Comment:
this is definitely no good, but it should be trivial to replace it with a lambda

Reply:
Yeah, I wasn't sure and originally had it exploded out in every instance. Unfortunately a lambda wouldn't work because I want to return from `DropItems()` if the dropcount is the treasure pool size. It is easy enough to explode it out though.

Reply:
the lambda can just return a bool and you can do `if (lambda()) return;`

Reply:
oh sorry, I don't see edits in email, but yeah you got it

Reply:
Not sure if you got notified but I made those changes :)